sympy/simplify/cse_main.py

```diff
@@ -536,3 +536,101 @@ def cse(exprs, symbols=None, optimizations=None, postprocess=None,
+
+    The input sequence should support iteration, indexing, and slicing.
+
+    To use concurrently with multiple sequences, concatenate them separated by
+    markers (custom markers can be pass in through the 'ignore' parameter').
+
+    Examples
+    ========
+
+    In the examples here, we use strings. A more common example for SymPy
+    would be the args of a noncommutative Mul.
+
+    >>> from sympy.simplify.cse_main import shortest_repeated_subsequence
+    >>> s = 'abcabc'
+    >>> shortest_repeated_subsequence(s)
+    'abc'
+    >>> print(shortest_repeated_subsequence('abc'))
+    None
+
+    >>> from itertools import chain
+    >>> from sympy.utilities.iterables import flatten
+    >>> s1, s2, s3 = 'abc', 'babc', 'abca'
+    >>> marker = '$'
+    >>> S = ''.join(chain(*flatten(zip([s1, s2, s3], '$$$'))))[:-1]
+    >>> S
+    'abc$babc$abca'
+    >>> shortest_repeated_subsequence(S, ignore=['$'])
+    'abc'
+    """
+    from sympy import oo
+
+    # This is based on the longest_common_substring algorithm, for instance,
+    # at
+    # https://en.wikibooks.org/wiki/Algorithm_Implementation/Strings/Longest_common_substring#Python3. We
+    # build a matrix to keep track of common subsequence of S with itself. For instance, for
+    # the sequence [a, b, c, a, b], we would have
+
+    #        a  b  c  a  b
+    #   [[0, 0, 0, 0, 0, 0],
+    # a  [0, 0, 0, 0, 1, 0],
+    # b  [0, 0, 0, 0, 0, 2],
+    # c  [0, 0, 0, 0, 0, 0],
+    # a  [0, 1, 0, 0, 0, 0],
+    # b  [0, 0, 2, 0, 0, 0]]
+
+    # (the leading row and column of 0s are used so we can always reference
+    # the previous row and column). Each entry in the matrix represents the
+    # longest subsequence ending in those two entries. We additionally keep
+    # the diagonal zero, so that we don't consider the entire sequence to be a
+    # common subsequence with itself.
+
+    m = [[0] * (1 + len(S)) for i in range(1 + len(S))]
+
+    shortest, x_shortest = oo, 0
+    for x in range(1, 1 + len(S)):
+        if S[x-1] in ignore:
+            continue
+        for y in range(1, 1 + len(S)):
+            if x == y:
+                continue
+            if S[x - 1] == S[y - 1]:
+                m[x][y] = (m[x-1][y-1] + 1)
+
+    # Now we traverse the matrix from the bottom up, keeping track diagonal
+    # "runs" (representing the same subsequence), and look for the shortest
+    # one that is >= 2 in length. If there are ties, the earliest one will be
+    # returned, but it would be easy to extend this to return all shortest
+    # subsequences.
+
+    seen = {}
+    for row in reversed(m):
+        seen = {i-1:seen[i]-1 for i in seen if seen[i]}
+        for x in range(len(row)):
+            if row[x] >= 2 and x not in seen:
+                if row[x] < shortest:
+                    shortest = row[x]
+                    x_shortest = x
+                seen[x] = row[x]
+
+    if shortest == oo:
+        return None
+    return S[x_shortest - shortest: x_shortest]
```

Comment:
This is my attempt at it:

```python
def compact(s, marker='$'):
    """Return 
        a) the recursively identified repeated sequences in the
            list of values in s and 
        b) the values of s with those pairs replaced with
            backslash prefaced numbers

    `marker` must be a character not in any s value.

    Examples
    ========

    >>> compact('xx'.split())
    ([], ['xx'])
    >>> compact('xxx'.split())
    ([], ['xxx'])
    >>> compact('xxxx'.split())
    ([('\\2', 'xx')], ['\\2\\2'])
    >>> compact('abc abe fabcall'.split())
    ([('\\7', 'ab'), ('\\8', '\\7c')], ['\\8', '\\7e', 'f\\8all'])

    """
    from collections import defaultdict as D
    assert not any(marker in i for i in s)
    # remap symbols to numbers
    sym = set()
    s = marker.join(s)
    for i in s:
        sym.add(i)
    sym.discard(marker)
    NEXT = len(sym)
    map = dict(list(enumerate(sym)))
    # 0 is reserved so move it
    map[NEXT] = map.pop(0)
    # -1 is for the marker
    map[-1] = marker
    # {symbol: number}
    pam = dict([(v,k) for k,v in map.items()])
    # recode s to numbers
    s = [pam[i] for i in s]
    reps = []  # (NEXT: repeated pair) items
    d = D(list)
    # identify where pairs are and keep track of
    # which pair occurs most often
    M = 0
    for i in range(len(s)-1):
      k = tuple(s[i:i+2])
      if -1 not in k:
        d[k].append(i)
        M = max(M, len(d[k]))
    while 1:
        if M <= 1:
            break
        dud = []  # pairs to be removed
        new = D(list)  # pairs to be added
        z = [0]*len(s)  # numbers of zeros to the left of position
        for k in d:
            ok = []
            zap = None
            if len(d[k]) == M:
                for i in d[k]:
                    if all(o==n for o,n in zip(k, s[i:i+2])):
                        # no substitution has made this invalid
                        ok.append(i)
                        zap=i+1,s[i+1]
                        s[i+1]=0
                if len(ok) > 1:
                    NEXT += 1
                    reps.append((NEXT, k))
                    for i in ok:
                        # put in the NEXT number and a 0 to
                        # maintain the length of s
                        s[i:i+2] = [NEXT, 0]
                        for j in range(i+2,len(s)):
                            z[j] += 1
                        if i+2 < len(s) and s[i+2] > 0:
                            # record a new pattern
                            new[(NEXT, s[i+2])].append(i)
                elif zap:
                    i, v = zap
                    s[i] = v
                dud.append(k)
            elif len(d[k]) == 1:
                dud.append(k)
            else:
                pass
        # remove those that are not repeated
        while dud:
            del d[dud.pop()]
        # add the new pairs
        d.update([(k,v) for k, v in new.items() if len(v) > 1])
        # get rid of place-holding 0s
        s = filter(None, s)
        # update locations for zeros
        for k in d:
            d[k] = [i - z[i] for i in d[k]]
        # find length of most repeated pair
        M = 0
        for k in d:
          M = max(M, len(d[k]))
    def getter(i):
        return map.get(i, '\\%s'%i)
    for i in range(len(reps)):
        k,v = reps[i]
        reps[i] = (getter(k), ''.join(map.get(i, getter(i)) for i in v))
    s = ''.join([getter(i) for i in s])
    return reps, s.split(marker)
```

Of course s will be a list of nc Muls and iteration over them would be different, but I think I would still recast the SymPy objects into Python primitives (ints and lists).